### ISSUE TITLE
Read FileStore after application restart

### ISSUE THREAD

**Issue by FabianInostroza** [eclipse/paho.mqtt.golang#90]
The documentation of the FileStore says 

> FileStore implements the store interface using the filesystem to provide true persistence, even across client failure

but I could not find how to send the stored messages after the application restarts, could you please tell me how is it done?

**Comment by alsm** [comment]
The intention is that if you open a Store that has Outbound messages in it they should be automatically resent when the client connects, but it does not appear that the client currently does this.

**Comment by FabianInostroza** [comment]
Yes, I tried to read the messages in the file store to send them in my application but I don't know how to convert the type. I would fix this myself but I'm new to golang

**Comment by jnicholls** [comment]
This is a pretty big gap.  It would be really nice of the Store actually did what it advertises.  I would also suggest keeping the store Packet agnostic and make it more application message oriented, i.e. just store the bytes that the caller intended to send.  I don't think doing it at the Packet level is the right approach, but I could be missing something.  I'd store messages that were in-flight and not acknowledged, and remove them when acknowledged.  On a restart/restore, the messages are sent again from scratch (rather than trying to continue the packet-level exchange).  Thoughts?

**Comment by jnicholls** [comment]
Bump, interested in @alsm's thoughts.  I'd be happy to whip up a PR, but would like to understand the reasoning behind the packet-level store first.  I will readily admit I am not an expert on the MQTT protocol and would love to learn more from you.  Can resuming an exchange at the packet-level after a large period of time actually work?

I'm very curious how we can bring this commented line back into action: https://github.com/eclipse/paho.mqtt.golang/blob/master/client.go#L260

Thanks in advance!

**Comment by alsm** [comment]
Hi, It is a bit of an embarrassment that the client appears to offer something that isn't actually complete. You only need to store Publishes, so there isn't too much of an issue with the store being packet specific, and the issue with it being agnostic is you still need to store the packet metadata such as QoS and Retain flags.
The only issue with absolutely resending every stored message is in the case of a QoS2 message where you have received the PUBREC from the server, in that case you would only need to send the PUBREL. To resend the original PUBLISH with the same packet id is a protocol violation according to the [mqtt 3.1.1 spec section 4.3.3](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718102).
Resuming after a long period of disconnection depends on the configuration of the server as to whether it will continue to store the session state, the CONNACK packet has a SessionPresent flag that will indicate this.
If you'd like to start a PR for this I'd be happy to help and review it.

**Comment by jnicholls** [comment]
@alsm Thanks for the information!
